add documentation for testing API (#260)
denoland/manual: language_server/testing_api.md

## Testing API

The Deno language server supports a custom set of APIs to enable testing. These

> are built around providing information to enabled the

Reviewer: ```suggestion
are built around providing information to enable the
```

[vscode's Testing API](https://code.visualstudio.com/api/extension-guides/testing)
but can be used by other language server clients to provide a similar interface.

### Capabilities

Both the client and the server should support the experimental `testingApi`
capability:

```ts
interface ClientCapabilities {
  experimental?: {
    testingApi: boolean;
  };
}
```

```ts
interface ServerCapabilities {
  experimental?: {
    testingApi: boolean;
  };
}
```

When a version of Deno that supports the testing API encounters a client which
supports the capability, it will initialize the code which handles the test
detection and will start providing the notifications which enable it.

It should also be noted that when the testing API capabilities are enabled, the
testing code lenses will no longer be sent to the client.

### Settings

> There are specific settings which change the behavior language server:

Reviewer: ```suggestion
There are specific settings which change the behavior of the language server:
```

- `deno.testing.args` - An array of strings which will be provided as arguments

>   when executing tests. Working in the same fashion as the `deno test`

Reviewer: ```suggestion
  when executing tests. This works in the same fashion as the `deno test`
```

  subcommand.

> - `deno.testing.enable` - a binary flag that enables or disables the testing

Reviewer: ```suggestion
- `deno.testing.enable` - A binary flag that enables or disables the testing
```

Or other one above lower case.

}

interface TestPassed {
  /** The state change that has occurred to a specific test or test step. */
  type: "passed";

  /** The test or test step relating to the state change. */
  test: TestIdentifier;

  /** An optional duration, in milliseconds from the start to the current
   * state. */
  duration?: number;
}

interface TestOutput {
  /** The test or test step has output information / logged information. */
  type: "output";

  /** The value of the output. */
  value: string;

  /** The associated test or test step if there was one. */
  test?: TestIdentifier;

  /** An optional location associated with the output. */
  location?: Location;
}

interface TestEnd {
  /** The test run has ended. */
  type: "end";
}

type TestRunProgressMessage =
  | TestEnqueuedStartedSkipped
  | TestFailedErrored
  | TestPassed
  | TestOutput
  | TestEnd;

interface TestRunProgressParams {
  /** The test run ID that the progress message applies to. */
  id: number;

  /** The message*/
  message: TestRunProgressMessage;
}
```

### Requests

The server handles two different requests:

#### `deno/testRun`

To request the language server to perform a set of tests, the client sends a

> `deno/testRun` request, which includes that ID of the test run to be used in

Reviewer: ```suggestion
`deno/testRun` request, which includes the ID of the test run to be used in
```
